Improve code generation for guards

bjorng opened the pull request:
There is a currently a compiler pass called `beam_bool` that tries to optimize the code generated for guards. It does not always succeed.

This pull request improves the code generation for guards by adding optimization of the intermediate code for guards to the `v3_kernel` pass which is run before code generation, and it removes the `beam_bool` pass. See the commit messages and the comments in the code for more details.

I have tried to make sure that Core Erlang generated by other front ends to the compiler (such as LFE) should still be correctly handled, but I could have missed something. Elixir should probably work fine, since it generates abstract code, not Core Erlang.

I am interested hearing about any code that breaks when compiled with this compiler, whether it is compiler crashed or incorrect behavior of the compiled code.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

psyeugenic commented:
Nice work!

I have a couple of initial remarks. As you know, I do not like that we add specialized expressions in v3_kernel and v3_codegen. I feel that we should reduce the number of different types of expressions, wherever it is possible.

In this case you have added `#k_inverted_test{}` as a specialized test expression. I understand why you want to do it. By design there is no easy way to swap the Fail and Success labels in kernel. What do you think about changing the existing `#k_test{}` expression instead? For instance by adding an argument to it, e.g. `#k_test{inverted :: bool()}`? Then let v3_codegen swap Success/Fail labels depending on the value of this argument. You still have to change the life pass though by passing an extra argument. Perhaps marginally better.

Nitpicking regarding the commit "beam_dead: Remove redundant 'bif' instructions"
I think the commit comment is a bit misleading or at least unclear. When I read it I interpreted "it" as referring to the jump instruction and not the bif/gc_bif instructions. Also, the code feels a bit construed by using `To =:= BifFail` instead of checking this with matching in the function head.

bjorng commented:
Thanks for the suggestion.

I have pushed a `fixup` commit that eliminates `#k_inverted_test{}` in favor of an extended `#k_test{}`. As you can see, I could eliminate the extra clauses in `extract_vars/2` and `uexpr/3` in `v3_kernel`. I could also eliminate a clause in `v3_codegen:bsm_rename_ctx/4`; a clause that was not covered.

bjorng commented:
I have pushed a `squash` commit with a clearer commit message.

The `To =:= BifFail` construction was my attempt to cover up an uncovered line. :-)

It turns out that there is no need to test whether the destination register for the `bif` or `gc_bif` instruction will be used. I found that out when `beam_validator` slapped me on my fingers when I tried to write a test case in BEAM assembly code to improve the coverage. So I have removed the liveness test of the destination register. See the new commit message and comments in the code for more details.

psyeugenic commented:
Is it necessary to keep the LFE source code around? It's not used for anything. If you want to keep them, it would be nice to know how to regenerate the .core files from the source.

bjorng commented:
> Is it necessary to keep the LFE source code around? It's not used for anything. If you want to keep them, it would be nice to know how to regenerate the .core files from the source.

No, it is not necessary. I just included them for reference. I suppose it would be better to just point to the LFE repository on github in my commit message.

psyeugenic commented:
Question, should this be optimized?
```erlang
inverted_guard(A,B,C,D) when (not (A < B)) andalso (not (B < C)) 
                                           andalso (not is_list(D)) -> ok.
```
This will generate:
```erlang
{function, inverted_guard, 4, 2}.
  {label,1}.
    {line,[{location,"t.erl",14}]}.
    {func_info,{atom,t},{atom,inverted_guard},4}.
  {label,2}.
    {bif,'<',{f,1},[{x,0},{x,1}],{x,4}}.
    {test,is_ne_exact,{f,1},[{x,4},{atom,true}]}.
    {test,is_ge,{f,1},[{x,1},{x,2}]}.
    {bif,is_list,{f,1},[{x,3}],{x,4}}.
    {bif,'not',{f,1},[{x,4}],{x,4}}.
    {test,is_eq_exact,{f,1},[{x,4},{atom,true}]}.
    {move,{atom,ok},{x,0}}.
    return.
```
i.e. we have `bif` instead of `test` instructions.

bjorng commented:
I'll have a look.

bjorng commented:
> Is it necessary to keep the LFE source code around? It's not used for anything. If you want to keep them, it would be nice to know how to regenerate the .core files from the source.

I have removed the LFE source code and added instructions on how to generate the .core files from the LFE repository.

bjorng commented:
>Question, should this be optimized?

```

inverted_guard(A,B,C,D) when (not (A < B)) andalso (not (B < C)) 
                                           andalso (not is_list(D)) -> ok.

```

This code pointed to a bigger problems.

I have now made the optimizations attempted in `v3_kernel` more conservative. In particular, I don't try to handle `#k_guard{}`.

To somewhat compensate, some optimizations in `sys_core_fold` are no longer done for code in guards. The optimizations would thwart the guard optimizations in `v3_kernel` (especially the optimizations that tried to eliminate `not`).

Your example will now be optimized (it is included in `guard_SUITE` as `complex_not_1b/1`).

psyeugenic commented on lib/compiler/src/v3_kernel.erl:
```diff
@@ -190,9 +190,487 @@ body(Ce, Sub, St0) ->
+    throw(not_possible);
+rewrite_bif('not', [Arg], Forest0, Inv, St) ->
+    {Expr,Forest} = forest_take_expr(Arg, Forest0),
+    case extract_bif(Expr) of
```
Can't we call `rewrite_bool` directly instead?

bjorng replied:
Yes, we can.

psyeugenic commented on lib/compiler/src/v3_kernel.erl:
```diff
@@ -190,9 +190,487 @@ body(Ce, Sub, St0) ->
+%%  forest_take_expr/2.
+
+forest_pre_seq(Exprs, Forest) ->
+    Es0 = [#k_var{name=V} || {unref,V} <- maps:keys(Forest)],
```
`unref` is not very informative. Maybe `untaken` would be better?

psyeugenic commented on lib/compiler/src/v3_kernel.erl:
```diff
@@ -190,9 +190,487 @@ body(Ce, Sub, St0) ->
+rewrite_match_select_1(_V, _, _, _Forest, _Inv, _St) ->
+    throw(not_possible).
+
+always_returns_bool(V, Forest) ->
```
Confusing name? `is_bool_var_rewritable`?

bjorng replied:
I'll rename it to `is_bool_expr` and add some comments.

psyeugenic commented on lib/compiler/src/v3_kernel.erl:
```diff
@@ -190,9 +190,487 @@ body(Ce, Sub, St0) ->
+    throw(not_possible).
+
+rewrite_match_select_1(Var, False, #k_atom{val=true}, Forest0, Inv, St0) ->
+    case False of
```
Maybe an "orelse" comment here would be useful?

bjorng replied:
Will add.

psyeugenic commented on lib/compiler/src/v3_kernel.erl:
```diff
@@ -190,9 +190,487 @@ body(Ce, Sub, St0) ->
+
+will_fail(#k_var{}) -> true;
+will_fail(#k_atom{val=false}) -> true;
+will_fail(#k_call{}) -> true;
```
Lazy! check error call! =)

bjorng replied:
Turns out that I don't need to check for the call to `error/1` -- the rewriting of the first expression will always fail. I can simplify the handling of #k_match{} without losing any optimizations.
